cmd/skopeo/signing.go

```diff
@@ -88,3 +89,137 @@ var standaloneVerifyCmd = cli.Command{
 	ArgsUsage: "MANIFEST DOCKER-REFERENCE KEY-FINGERPRINT SIGNATURE",
 	Action:    standaloneVerify,
 }
+
+func addSignature(context *cli.Context) error {
+	if len(context.Args()) != 2 {
+		return errors.New("Usage: skopeo add-signature reference key-fingerprint")
+	}
+	ref := context.Args()[0]
+	fingerprint := context.Args()[1]
+
+	imageRef, err := transports.ParseImageName(ref)
+	if err != nil {
+		return fmt.Errorf("Error parsing %s: %v", ref, err)
+	}
+	name := transports.ImageName(imageRef)
+
+	// TODO: the signed named reference is always relative to the transport.  Is this the right value?
+	signReference := imageRef.StringWithinTransport()
+
+	systemContext := contextFromGlobalOptions(context)
+	src, err := imageRef.NewImageSource(systemContext, nil)
+	if err != nil {
+		return fmt.Errorf("Error reading %s: %v", name, err)
+	}
+	defer src.Close()
+
+	manifest, _, err := src.GetManifest()
+	if err != nil {
+		return fmt.Errorf("Error reading manifest for %s: %v", name, err)
+	}
+
+	mech, err := signature.NewGPGSigningMechanism()
+	if err != nil {
+		return fmt.Errorf("Error initializing GPG: %v", err)
+	}
+
+	signature, err := signature.SignDockerManifest(manifest, signReference, mech, fingerprint)
+	if err != nil {
+		return fmt.Errorf("Error creating signature: %v", err)
+	}
+
+	dest, err := imageRef.NewImageDestination(systemContext)
+	if err != nil {
+		return fmt.Errorf("Error writing to %s: %v", name, err)
+	}
+	defer dest.Close()
+
+	signatures, err := src.GetSignatures()
+	if err != nil {
+		return fmt.Errorf("Error reading signatures from %s: %v", name, err)
+	}
+
+	signatures = append(signatures, signature)
```

Comment:
This will keep adding signatures even if a signature already exists for the (fingerprint, `signReference`) combination. That of course doesn’t hurt, and we don’t easily provide a way to avoid that; I guess we could only explicitly document that we might do nothing in the future, if the required signature is already present.